pkg/cloudprovider/providers/aws/device_allocator.go

```diff
@@ -40,11 +45,27 @@ type DeviceAllocator interface {
 	// name. Only the device suffix is returned, e.g. "ba" for "/dev/xvdba".
 	// It's up to the called to add appropriate "/dev/sd" or "/dev/xvd" prefix.
 	GetNext(existingDevices ExistingDevices) (mountDevice, error)
+
+	// LockForUse will lock the device allocator, so as another goroutine can't use it
+	LockForUse()
+
+	// Mark for use makes last selected device as taken
+	MarkForUse()
```

Comment:
The problem is that I think that the two pools are heavily inter-dependent right now, and this makes the code fragile and difficult to review.  When we do an attach, we put an entry into the attaching map, take a lock on the device allocator and take the next entry from the device allocator (so nobody else can take it), attach the volume, mark the device in use in the device allocator and unlock it and remove the entry from the attaching map.  Even if you can get that right today, it's just too complicated to keep correct, particularly as we don't have e2e coverage on this.